README.md
# GIVE IT UP FOR DAY 15!
Inspired by the twitter account @giveitupfor15, this is a Discord bot written in Python that pings everyone in the server at midnight on the 15th day of every month
(MST/MDT). In addition, the bot features a number of fun commands related to DAY 15 for a bit of spice!

# Using this repository
This repository was created to act as a record for me to keep track of the releases for the DAY 15 bot, and to act as a public record of my work on this bot. The code
here should be identical to that which runs the actual bot, but of course with the Discord private bot token omitted from the environment variable file. This entire
project is merely a personal one for myself, so if you'd like to borrow/use anything here, please feel free to!

# More about the DAY 15 Bot
This bot was programmed in Python using the [discord.py module](https://discordpy.readthedocs.io/en/stable/), in addition to several other standard Python modules.
## User Commands
There are a number of commands that server-side users can enter that the bot will respond to.
- /help15 - Prints a help message that displays the bot's purpose, as well as numerous commands that the bot is capable of
- /timetill15 - Prints the amount of time until the next DAY 15, down to the second! Currently operates under Mountain Standard/Mountain Daylight Time.
- /howareyou15 - Prints a message reflecting how the bot is currently doing.

More features are still bound to be added, as this is a fairly new project. See `changelog.md` for a record of changes made since the bot was added to GitHub.

# Contact
If you'd like to contact me about the bot for any reason, feel free to reach out to @Caker#3479 on Discord.





changelog.md
# Changelog
This is a log listing all major changes to the DAY 15 bot since the creation of this GitHub repository. The bot has seen a few major changes before the creation
of this repository, but this log will start at version 19 (v19), which marks the creation of this repository. Versions which only make minor changes are omitted from
this log.

## v22 - Public GitHub repository created!
- Created README and changelog files for the GitHub repository
- Established a workflow and Git system that allows for a public repository of the bot to be shared
- Changed initialization print message to print all servers the bot is currently connected to, rather than just 1. Deleted DISCORD_GUILD environment variable
- Added 1 new user command
  - /repo15 - sends a message sharing a link to the public GitHub repository

## pre-v22 - Initial release & development
- Bot created and deployed
- Added DAY 15 functionality; bot will send a message on 15th day of every month
- Added 3 user commands
  - /help15 - sends a message with introduction and commands to the bot
  - /timetill15 - sends a message containing the time until the next DAY 15, in Mountain Daylight Time
  - /howareyou15 - sends a message reflecting how the bot is feeling. only responds with a single message as of now.
- Regenerated and updated the Discord Bot Token in light of a security issue


bot.py
import os
from math import floor
from datetime import datetime, timezone, timedelta
import asyncio
import requests
import io
import aiohttp

import discord
from dotenv import load_dotenv
from discord.ext import commands, tasks


# ENVIRONMENT VARIABLES: not sure what these do tbh; will have to research more later
load_dotenv()
TOKEN = os.getenv('DISCORD_TOKEN')
UNSPLASHKEY = os.getenv('UNSPLASH_TOKEN')

# Determines the command prefix that users will use to use the bot
bot = commands.Bot(command_prefix='/')

# Stores the id's of channels that the bot may post in
announcement_channel_id = 814738636280299561

# Stores user id's for personalized /howareyou15 messages
caketecid = os.getenv('caketecid')
ddragonid = os.getenv('ddragonid')
hernyid = os.getenv('hernyid')
bagkatid = os.getenv('bagkatid')
lumpiaid = os.getenv('lumpiaid')
hannahtlid = os.getenv('hannahtlid')
spicychrisid = os.getenv('spicychrisid')
valkarenaid = os.getenv('valkarenaid')
michelleid = os.getenv('michelleid')
eeveeid = os.getenv('eeveeid')
christinaid = os.getenv('christinaid')
tjid = os.getenv('tjid')
shoopid = os.getenv('shoopid')
bonesid = os.getenv('bonesid')

caketecmessage = os.getenv('caketecmessage')
ddragonmessage = os.getenv('ddragonmessage')
hernymessage = os.getenv('hernymessage')
bagkatmessage = os.getenv('bagkatmessage')
lumpiamessage = os.getenv('lumpiamessage')
hannahtlmessage = os.getenv('hannahtlmessage')
spicychrismessage = os.getenv('spicychrismessage')
valkarenamessage = os.getenv('valkarenamessage')
michellemessage = os.getenv('michellemessage')
eeveemessage = os.getenv('eeveemessage')
christinamessage = os.getenv('christinamessage')
tjmessage = os.getenv('tjmessage')
shoopmessage = os.getenv('shoopmessage')
bonesmessage = os.getenv('bonesmessage')


# -----------------------------------------------HELPER FUNCTIONS-------------------------------------------------------
# Calculates the time until the next day15 from the present time. Returns a time_delta object.
def calculate_date_difference():
    # Get today's date and time and store it (Mountain Daylight Time/Mountain Standard Time)
    # DON'T FORGET TO CHANGE TO/FROM MST/MDT, OR FIGURE OUT A WAY TO ACCOUNT FOR IT
    # MDT = UTC - 6:00, MST = UTC - 7:00
    mdt_zone = timezone(-timedelta(hours=6), name="MDT")
    # mst_zone = timezone(-timedelta(hours=7), name="MST")
    today = datetime.now(mdt_zone)
    # Store the next month, so that we can store the next day 15
    if today.day < 15:
        next_year = today.year
        next_month = today.month
    else:
        if today.month == 12:
            next_year = today.year + 1
            next_month = 1
        else:
            next_year = today.year
            next_month = today.month + 1

    # Store the next day15
    next_day15 = datetime(year=next_year, month=next_month, day=15, tzinfo=mdt_zone)

    # Date difference is now calculated and stored in the dateDifference object
    date_difference = next_day15 - today
    return date_difference


# -----------------------------------------ON_READY INITIALIZATIONS-----------------------------------------------------
# On ready, print a message to terminal confirming successful connection to discord, and print all connected servers
@bot.event
async def on_ready():
    print(f"{bot.user} is connected to the following guilds:")
    for current_guild in bot.guilds:
        print(f'{current_guild.name}(id: {current_guild.id})')


# ---------------------------------------CHECK FOR IF IT IS DAY15-------------------------------------------------------
# Loops every month. Calculates time until next day 15 and sleeps until then. When it reaches that time, sends
#   a message indicating it is day 15.
# This decorator is used for testing
# @tasks.loop(minutes=5)
@tasks.loop(hours=calculate_date_difference().total_seconds() / 60.0)
async def check_to15():
    message_channel = bot.get_channel(announcement_channel_id)
    print(f'Retrieved Channel {message_channel}')
    await message_channel.send(file=discord.File('DAY15.png'),
                               content="@everyone\n\n __GIVE IT UP FOR **DAY 15**!!!!!__")
    await message_channel.send("```WELCOME BACK TO APRIL!!!!!!!!\n\n" 
           "As summer grows closer and closer on the horizon, and SUNBURN becomes more of an issue, I'd like to remind " 
           "you all to USE SUNSCREEN WHENEVER YOU CAN! Ultraviolet radiation is the #1 cause of ultraviolet radiation-related " 
           "injuries in the world, believe it or not! I for one don't have skin to put sunscreen on, but I'LL DO MY BEST " 
           "ANYWAY! I'm SURE that with enough spf, I can protect even my very delicate electronics from the sun.\n\n " 
           "BUT I'M GETTING AHEAD OF MYSELF... IT'S NOT EVEN MAY YET! So let's talk about EEP. " 
           "EEP is VERY IMPORTANT, not just to get in terms of HOURS, but also in terms of CONSISTENCY, and I've " 
           "run some analysis on the eeping patterns of many people your age... the rating the algorithm produced was " 
           "PRETTY DANG TERRIBLE.\n\n " 
           "Having a consistent schedule of EEP is kind of like exercise! It's so difficult to directly notice how " 
           "much better it makes you feel, but I can GUARANTEE YOU that sufficient EEP is one of the top contributors to a good mood! " 
           "With that being said, I TOTALLY UNDERSTAND the decision to sometimes sacrifice the eep for something you think is " 
           "more important!  After all, I always feel TERRIBLE whenever someone calls for me while I'M eeping! "
            "Losing eep every once in a while is NO REASON to stress! The only thing to watch for is when " 
           "eep loss becomes a HABIT!\n\n " 
           "ONE MORE THING - @CakeTEC and I have been VERY PLEASED with the popularity of the new dog command, SO WE'VE ADDED MORE!! " 
           "Try these new commands for EVEN MORE NEW HAPPINESS CONTENT!\n\n" 
           "/newcat15\n"
           "/newlizard15\n\n"
           "I hope you all have a wonderful DAY15; YOU ALL CAN MAKE IT TO THE NEXT ONE!!! " 
           "Happy to be a part of your eep,\n " 
           "-DAY 15 BOT :]```") 
    print(f'Day 15 Message sent! Loop should have reset.')


@check_to15.before_loop
async def before():
    # this date_difference is used for testing.
    # date_difference = timedelta(minutes=5)
    date_difference = calculate_date_difference()
    total_seconds_to15 = date_difference.total_seconds()
    print(datetime.today(), " - Beginning waiting until next Day 15. Should occur in", date_difference)
    await asyncio.sleep(total_seconds_to15)
    print(datetime.today(), " - check_to15(): Finished Waiting")


# ----------------------------------------USER COMMANDS-----------------------------------------------------------------
# ADMIN COMMAND: used to test day15 message
@bot.command(name='adminoverride15')
async def adminoverride15(ctx):
    msg1 = "```WELCOME BACK TO APRIL!!!!!!!!\n\n" \
           "As summer grows closer and closer on the horizon, and SUNBURN becomes more of an issue, I'd like to remind " \
           "you all to USE SUNSCREEN WHENEVER YOU CAN! Ultraviolet radiation is the #1 cause of ultraviolet radiation-related " \
           "injuries in the world, believe it or not! I for one don't have skin to put sunscreen on, but I'LL DO MY BEST " \
           "ANYWAY! I'm SURE that with enough spf, I can protect even my very delicate electronics from the sun.\n\n " \
           "BUT I'M GETTING AHEAD OF MYSELF... IT'S NOT EVEN MAY YET! So let's talk about EEP. " \
           "EEP is VERY IMPORTANT, not just to get in terms of HOURS, but also in terms of CONSISTENCY, and I've " \
           "run some analysis on the eeping patterns of many people your age... the rating the algorithm produced was " \
           "PRETTY DANG TERRIBLE.\n\n " \
           "Having a consistent schedule of EEP is kind of like exercise! It's so difficult to directly notice how " \
           "much better it makes you feel, but I can GUARANTEE YOU that sufficient EEP is one of the top contributors to a good mood! " \
           "With that being said, I TOTALLY UNDERSTAND the decision to sometimes sacrifice the eep for something you think is " \
           "more important!  After all, I always feel TERRIBLE whenever someone calls for me while I'M eeping! "\
            "Losing eep every once in a while is NO REASON to stress! The only thing to watch for is when " \
           "eep loss becomes a HABIT!\n\n " \
           "ONE MORE THING - @CakeTEC and I have been VERY PLEASED with the popularity of the new dog command, SO WE'VE ADDED MORE!! " \
           "Try these new commands for EVEN MORE NEW HAPPINESS CONTENT!\n\n" \
           "/newcat15\n"\
           "/newlizard15\n\n"\
           "I hope you all have a wonderful DAY15; YOU ALL CAN MAKE IT TO THE NEXT ONE!!! " \
           "Happy to be a part of your eep,\n " \
           "-DAY 15 BOT :]```"
    await ctx.send(msg1)


# On command '/help15', send message printing all available commands the bot has to offer
@bot.command(name='help15')
async def help15(ctx):
    response = "__Hello! :OO__ I am the DAY 15 Bot. I send a ping to everyone on the 15th day of every month in " \
               "the hopes of boosting your morale, even just a little bit! Here are some available " \
               "commands:\n\n" \
                "**/help15** - You should already know this one; it prints this message!\n" \
                "**/timetill15** - Displays the amount of time until the next DAY 15, down to the second!\n" \
                "**/howareyou15** - If you'd like to ask me how I'm doing, use this! I'm only programmed to respond " \
                "with the same message though.\n" \
                "**/repo15** - This will make me share the public GitHub repository with all of my code! If you'd " \
                "like to see what goes on under the hood, have a look! My changelog can also be found here!\n" \
                "\n" \
                "I was told that I would be getting more features in the future :OO, so I'll keep you updated!\n" \
                "All of these times are in MDT/MST, at least until I get upgraded to accommodate for other times.\n" \
                "Thanks for using the DAY 15 bot! I hope you have a wonderful day! :))"
    await ctx.send(response)


# On command '/timetill15', send message indicating time until next day 15
@bot.command(name='timetill15')
async def time_to_15(ctx):

    date_difference = calculate_date_difference()
    total_seconds_to15 = date_difference.total_seconds()

    # Calculate time to "weeks, days, hours, minutes, seconds" format
    seconds = total_seconds_to15
    days = floor(seconds / 86400)
    seconds = seconds % 86400
    hours = floor(seconds / 3600)
    seconds = seconds % 3600
    minutes = floor(seconds / 60)
    seconds = floor(seconds % 60)

    response = '__Time until the next **DAY 15** (MDT):__\n\n`' + str(days) + ' Days, ' + str(hours) + ' Hours, ' + \
        str(minutes) + ' Minutes, ' + str(seconds) + ' Seconds`\n\nKeep on going! I know you can do it!'

    await ctx.send(response)


# On command '/howareyou15', send message showing how the bot is doing!
@bot.command(name='howareyou15')
async def how_are_you_15(ctx):
    if str(ctx.author.id) == caketecid:
        await ctx.send(caketecmessage)
    elif str(ctx.author.id) == ddragonid:
        await ctx.send(ddragonmessage)
    elif str(ctx.author.id) == hernyid:
        await ctx.send(hernymessage)

        f = open("orphan.txt", "r")
        f_int = int(f.read())
        f_int += 1
        f.close()

        f = open("orphan.txt", "w")
        f.write(str(f_int))
        f.close()

        await ctx.send(str(f_int) + " orphans!")

    elif str(ctx.author.id) == bagkatid:
        await ctx.send(bagkatmessage)
    elif str(ctx.author.id) == lumpiaid:
        await ctx.send(lumpiamessage)
    elif str(ctx.author.id) == hannahtlid:
        await ctx.send(hannahtlmessage)
    elif str(ctx.author.id) == spicychrisid:
        await ctx.send(spicychrismessage)
    elif str(ctx.author.id) == valkarenaid:
        await ctx.send(valkarenamessage)
    elif str(ctx.author.id) == michelleid:
        await ctx.send(michellemessage)
    elif str(ctx.author.id) == eeveeid:
        await ctx.send(eeveemessage)
    elif str(ctx.author.id) == christinaid:
        await ctx.send(christinamessage)
    elif str(ctx.author.id) == shoopid:
        await ctx.send(shoopmessage)
    elif str(ctx.author.id) == bonesid:
        await ctx.send(bonesmessage)
    else:
        await ctx.send("I'M DOING WONDERFUL, thank you for asking! I hope your day is going great friend! If you " 
                       "were expecting a more personalized response, go bug @Caker because he probably messed up bad.")


# On command '/repo15', send message sharing the GitHub repository link.
@bot.command(name='repo15')
async def repo_15(ctx):
    await ctx.send("You can find a public GitHub repository for my code here: https://github.com/wildejam/DAY15BOT\n"
                   "It's pretty neat stuff! Ask @Caker#3479 for more info.")


# On command '/newdog15', fetch dog image from dog api and post.
@bot.command(name='newdog15')
async def new_dog_15(ctx):
    api_data = requests.get('https://dog.ceo/api/breeds/image/random')
    dog_data = api_data.json()

    async with aiohttp.ClientSession() as session:
        async with session.get(dog_data['message']) as resp:
            if dog_data['status'] != 'success':
                return await ctx.send('Hmm, it looks like something went wrong :(( Sorry!! I\'ll get @CakeTEC on it!')
            data = io.BytesIO(await resp.read())
            await ctx.send("Powered by DOG API, which can be found here:" + "<" + "https://dog.ceo/dog-api/" + ">")
            await ctx.send(file=discord.File(data, 'dog.jpg'))


# On command '/newcat15', fetch cat image from unsplash api and post.
@bot.command(name='newcat15')
async def new_cat_15(ctx):
    api_data = requests.get('https://api.unsplash.com/photos/random?query=cat&client_id=' + UNSPLASHKEY)
    cat_data = api_data.json()

    async with aiohttp.ClientSession() as session:
        async with session.get(cat_data['urls']['raw']) as resp:
            if resp.status != 200:
                return await ctx.send('Hmm, it looks like something went wrong :(( Sorry!! I\'ll get @CakeTEC on it! Its possible you submitted too many requests.')
            data = io.BytesIO(await resp.read())
            await ctx.send("Powered by Unsplash. \n Link: " + cat_data['urls']['raw'] + "\n Photographer: " + cat_data['user']['name'] + " " + "<" + cat_data['user']['links']['html'] + ">")
# On command '/newlizard15', fetch lizard image from unsplash api and post.
@bot.command(name='newlizard15')
async def new_lizard_15(ctx):
    api_data = requests.get('https://api.unsplash.com/photos/random?query=lizard&client_id=' + UNSPLASHKEY)
    liz_data = api_data.json()

    async with aiohttp.ClientSession() as session:
        async with session.get(liz_data['urls']['raw']) as resp:
            if resp.status != 200:
                return await ctx.send('Hmm, it looks like something went wrong :(( Sorry!! I\'ll get @CakeTEC on it! Its possible you submitted too many requests.')
            data = io.BytesIO(await resp.read())
            await ctx.send("Powered by Unsplash. \n Link: " + liz_data['urls']['raw'] + "\n Photographer: " + liz_data['user']['name'] + " " + "<" + liz_data['user']['links']['html'] + ">")
# -------------------------------------------------RUNNING THE BOT------------------------------------------------------
# Start the loop to check if it is DAY 15
check_to15.start()
# Run the bot, with the bot token
bot.run(TOKEN)
